File: mpi2/mandelbrot.f90
program mandelbrot

complex z,kappa
integer green,blue,i,j,k

integer   N,maxiter
parameter (N=2000,maxiter=1000)
real      x(0:N-1,0:N-1)
    
! ----------------------------------------------------------------

do i=0,N-1
    do j=0,N-1
        kappa = cmplx((4.0*(i-N/2)/N),(4.0*(j-N/2)/N))

        k=1
        z=kappa
        do while (abs(z).le.2 .and. k.lt.maxiter) 
            k=k+1
            z=z*z + kappa
        end do

        x(i,j)=log(real(k))/log(real(maxiter))
    end do
 
    if (mod(i+1,N/10).eq.0) write(6,*) 100*(i+1)/N,'% complete.'
end do

! ----------------------------------------------------------------

write(6,*) 'Writing mandelbrot.ppm'

open(7,file='mandelbrot.ppm',status='unknown')
write(7,100) 'P6', N, N, 255

do j=0,N-1
    do i=0,N-1
        if (x(i,j).lt.0.5) then
            green=2.0*x(i,j)*255
            write(7,110) achar(255-green),achar(green),achar(0)
        else 
            blue= 2.0*x(i,j)*255 - 255
            write(7,110) achar(0),achar(255-blue),achar(blue)
        end if
    end do
end do

100  format(A2,/,I4,I5,/,I3)
110  format(3A1,$)

close(7)
end

File: mpi2/mandelbrot.c

#define MAXITER 1000
#define N	2000

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>

/* ----------------------------------------------------------------*/

int main() {
    int	   i,j,k,tenth;
    short  green,blue;
    float  **x;
    FILE   *fp;

    float complex   z, kappa;

    // Allocate x as a 2-D array - requires use of a double pointer
    x=(float**)malloc(N*sizeof(float*));
    for (i=0;i<N;i++) {
      x[i]=(float*)malloc(N*sizeof(float));
    }
  
    // Iterate over 2-D grid: i covers N cols and j covers N rows
    for (i=0; i<N; i++) {
        for (j=0; j<N; j++) {
	    
            //kappa is the coordinate in the complex plane of the (i,j)th pixel
            //both the real and imaginary axis range from -2 to 2 (4 units)
            kappa = (4.0*(i-N/2))/N + (4.0*(j-N/2))/N * I;

            //z_0 = kappa, initial value of z before iteration begins
            z = kappa;
	
            //now we iterate until either the max number of iterations is reached 
            //OR z leaves the circle of radius 2 centred at the origin
	    k=1; 
	    while ((cabs(z)<=2) && (k++<MAXITER)) {
	        z = z*z + kappa;
            }
	  
            // Store the number of iterations required for this point to leave the mandelbrot set.
            // If this point IS IN the Mandelbrot set, then k=MAXITER.
            // We actually store log(k)/log(MAXITER) to see more colour variation in the points 
            // outside the Mandelbrot set.
            // Points that are in the set have k=MAXITER and therefore log(k)/log(MAXITER)=1.0.
            // When the mandelbrot.ppm file is created below, x=1.0 translates to 100% blue.
	    x[i][j]=log((float)k) / log((float)MAXITER);
        }

        tenth=(int) N/10;
        if ((i+1)%tenth==0) 
            printf("%d%% complete.\n",100*(i+1)/N);
    }

/* ----------------------------------------------------------------*/
  
    printf("Writing mandelbrot.ppm\n");
    fp = fopen ("mandelbrot.ppm", "w");
    fprintf (fp, "P6\n%4d %4d\n255\n", N, N);
    
    for (j=0; j<N; j++)
        for (i=0; i<N; i++) 
	    if (x[i][j]<0.5) {
	        green= (int) (2*x[i][j]*255);
                fprintf (fp, "%c%c%c", 255-green,green,0);
	    } else {
	        blue= (int) (2*x[i][j]*255-255);
                fprintf (fp, "%c%c%c", 0,255-blue,blue);
	    }
    
    fclose(fp);

/* ----------------------------------------------------------------*/

    for (i=0; i<N; i++)
        free(x[i]);
    free(x);
}
